### PR TITLE
[SYSTEMML-1883] Add XOR built in operator

### PR DESCRIPTION
**Pull request by j143**
- [x] Adds a builtin `XOR` operator.
- [x] Corresponding tests
- [x] adds `xor` to pydef
- [ ] gpu version

 This operator is a requirement for `Bayesian.dml`

### CODE REVIEW

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2046/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2047/

**Comment by j143** [comment]
cc @nakul02 @niketanpansare 
can you please review this?. Please also comment on GPU operator requirement. 
```
05:49:45 testXor(org.apache.sysml.test.integration.functions.binary.scalar.XorTest)  Time elapsed: 2.839 sec  <<< FAILURE!
05:49:45 java.lang.AssertionError: unable to read file: File target/testTemp/functions/binary/scalar/XorTest/out/left_1 does not exist
05:49:45 	at org.junit.Assert.fail(Assert.java:88)
05:49:45 	at org.apache.sysml.test.utils.TestUtils.compareDMLMatrixWithJavaMatrix(TestUtils.java:355)
05:49:45 	at org.apache.sysml.test.integration.AutomatedTestBase.compareResults(AutomatedTestBase.java:1344)
05:49:45 	at org.apache.sysml.test.integration.AutomatedTestBase.compareResults(AutomatedTestBase.java:1296)
05:49:45 	at org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor(XorTest.java:56)
```

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2049/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2049/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2049/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor</strong></a></li></ul>

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2063/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2063/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2063/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor</strong></a></li></ul>

**Comment by j143** [comment]
Hi, some volunteer would like to review this PR. I believe this will be useful [here](https://github.com/apache/systemml/pull/632/files#diff-b8cc626bab85b5d70a648cf6394ac5ceR168)
perhaps cc @nakul02  @deroneriksson

**Comment by nakul02** [comment]
Try fixing the test case that is failing.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2114/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2115/

**Comment by j143** [comment]
Hi @nakul02 , 
the `XorTest.java` is still failing, due to some reason can you please have a pass over the change set?. Thanks.

File: src/main/java/org/apache/sysml/parser/dml/Dml.g4 [hunk 1]
```diff
@@ -152,6 +152,8 @@ expression returns [ org.apache.sysml.parser.common.ExpressionInfo info ]
     | left=expression op=('&'|'&&') right=expression # BooleanAndExpression
     // boolean or
     | left=expression op=('|'|'||') right=expression # BooleanOrExpression
+    // boolean xor
+    | op='xor' '(' left=expression ','  right=expression ')' # BooleanXorExpression
```
Comment (niketanpansare) [inline src/main/java/org/apache/sysml/parser/dml/Dml.g4]: I would highly recommend not adding `xor` as keyword in our grammar. Instead please consider adding xor as a builtin binary function. Most of your code will remain same, except for the language layer. You can use 
 https://github.com/apache/systemml/commit/50a895f86167d5766b249a3db89f9c2b06f7a0d0 as an example for adding a builtin at a language layer :)

Here is a sample R script that demonstrates overriding of builtin function:
```
> xor(1, 0)
[1] TRUE
> xor <- function(a, b) { a + b }
> xor(1, 0)
[1] 1
```

Here is a sample DML script that demonstrates overriding of builtin function:
```
$ cat temp.dml
max <- function(double a, double b) return (double c)  {
        c = a + b
}

d = max(1, 2)
print(d)
$ ~/spark-2.2.0-bin-hadoop2.7/bin/spark-submit SystemML.jar -f temp.dml
17/09/21 10:20:48 INFO DMLScript: BEGIN DML run 09/21/2017 10:20:48
17/09/21 10:20:48 INFO DMLConfig: Using internal default configuration settings.  If you wish to customize any settings, please supply a `SystemML-config.xml` file.
17/09/21 10:20:48 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
3.0
SystemML Statistics:
Total execution time:           0.000 sec.
Number of executed Spark inst:  0.
```

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2116/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2116/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2116/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor</strong></a></li></ul>

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2123/<h2>Failed Tests: <span class='status-failure'>5</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2123/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>5</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2123/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2123/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.gdfo/GDFOLinregDS/testGDFOLinregDSSparseMR/'><strong>org.apache.sysml.test.integration.functions.gdfo.GDFOLinregDS.testGDFOLinregDSSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2123/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.gdfo/GDFOLinregDS/testGDFOLinregDSDenseMR/'><strong>org.apache.sysml.test.integration.functions.gdfo.GDFOLinregDS.testGDFOLinregDSDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2123/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.gdfo/GDFOLinregDSsimpl/testGDFOLinregDSSparseMR/'><strong>org.apache.sysml.test.integration.functions.gdfo.GDFOLinregDSsimpl.testGDFOLinregDSSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2123/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.gdfo/GDFOLinregDSsimpl/testGDFOLinregDSDenseMR/'><strong>org.apache.sysml.test.integration.functions.gdfo.GDFOLinregDSsimpl.testGDFOLinregDSDenseMR</strong></a></li></ul>

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2129/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2132/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2133/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2135/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2135/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2135/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor</strong></a></li></ul>

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2136/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2136/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2136/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor</strong></a></li></ul>

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2138/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2138/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2138/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor</strong></a></li></ul>

**Comment by j143** [comment]
Niketan, I am missing out something here. 😕

File: src/main/java/org/apache/sysml/hops/rewrite/RewriteConstantFolding.java [hunk 2]
```diff
@@ -131,7 +131,11 @@ else if( isApplicableTrueDisjunctivePredicate(root) )
 		{
 			literal = new LiteralOp(true);
 		}
-		
+		//fold biconditional predicate if both the inputs are same
+		else if(isApplicableFalseBiconditionalPredicate(root) )
```
Comment (j143) [inline src/main/java/org/apache/sysml/hops/rewrite/RewriteConstantFolding.java]: the name `biconditional predicate` may or may not be appropriate. But, the logic seems correct.

File: src/main/cpp/kernels/SystemML.cu [hunk 3]
```diff
@@ -106,6 +106,7 @@ __forceinline__ __device__ double binaryOp(double x, double y, int op) {
         case 12 : return max(x, y);
         case 13 : return getBoolean((int)llrint(x) & (int)llrint(y));
         case 14 : return getBoolean((int)llrint(x) | (int)llrint(y));
+	case 19 : return getBoolean(x ^ y);
```
Comment (j143) [inline src/main/cpp/kernels/SystemML.cu]: `llvm` installation failed due to system hardware check. So, with my next task for implementing matrix operations for `xor, !, |` I will change this file, for now will remove the change.

**Comment by mboehm7** [comment]
I guess the test is failing because the output names are not set correctly. Btw, this is a very old (+5 years) way of running test scripts. Unfortunately, there are still some of these instances in our test suite. You can use `FullStringComparisonTest` or most of the matrix tests as a template. This makes the input/output handling very explicit and should easily allow us to resolve this issue.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2142/

**Comment by j143** [comment]
Test this Jenkins

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2143/

**Comment by j143** [comment]
Test this Jenkins

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2145/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2145/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2145/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor</strong></a></li></ul>

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2150/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2150/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2150/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor</strong></a></li></ul>

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/<h2>Failed Tests: <span class='status-failure'>8</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>8</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXorLeft1/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXorLeft1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXorLeft2/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXorLeft2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXorLeft3/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXorLeft3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXorLeft4/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXorLeft4</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXorRight1/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXorRight1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXorRight2/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXorRight2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXorRight3/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXorRight3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2151/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXorRight4/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXorRight4</strong></a></li></ul>

**Comment by j143** [comment]
Hi @niketanpansare , with @mboehm7 's suggestion I changed to new testing. Can you please have another pass. Thanks.

**Comment by j143** [comment]
Once this is over, I will try moving to the matrix operation support.  :)

**Pull request closed without merging**